File: src/vct/col/ast/IfStatement.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import hre.ast.MessageOrigin;

import java.util.*;

public class IfStatement extends ASTNode {

  public IfStatement(){
  }
  
  public IfStatement(ASTNode cond,ASTNode true_branch,ASTNode false_branch){
    addClause(cond,true_branch);
    if(false_branch!=null){
      addClause(else_guard,false_branch);
    }
  }
  
  

  public static final ASTNode else_guard=new ConstantExpression(true,new MessageOrigin("else guard"));

  private static class Case {
    public ASTNode guard;
    public ASTNode  statement;
  }
  
  private ArrayList<Case> cases=new ArrayList();
  
  public void addClause(ASTNode guard,ASTNode s){
    Case clause=new Case();
    clause.guard=guard;
    clause.statement=s;
    cases.add(clause);
  }

  public int getCount(){ return cases.size(); }

  public ASTNode getGuard(int i){ return cases.get(i).guard; }
  
  public ASTNode getStatement(int i){ return cases.get(i).statement; }

  public void accept_simple(ASTVisitor visitor){
    visitor.visit(this);
  }

}

File: vercors/src/main/java/vct/col/ast/stmt/composite/IfStatement.scala
package vct.col.ast.stmt.composite

import hre.ast.MessageOrigin
import vct.col.ast.expr.constant.ConstantExpression
import vct.col.ast.generic.{ASTNode, DebugNode}
import vct.col.ast.stmt.composite.BlockStatement
import vct.col.ast.util.ASTVisitor

import scala.collection.mutable.ArrayBuffer
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

object IfStatement {
  val elseGuard = new ConstantExpression(true, new MessageOrigin("else guard"))
}

case class IfStatementCase(var guard:ASTNode, var effect:ASTNode) extends DebugNode {
  override def debugTreeChildrenFields: Iterable[String] = Seq("guard", "effect")
  override def debugTreePropertyFields: Iterable[String] = Seq()
}

class IfStatement extends ASTNode with VisitorHelper {
  private[this] val cases = new ArrayBuffer[IfStatementCase]()
  
  def getCount = cases.size
  def getGuard(i:Int) = cases.apply(i).guard
  def getStatement(i:Int) = cases.apply(i).effect
  
  def this(cond:ASTNode, truebranch:ASTNode, falsebranch:ASTNode) = {
    this()
    addClause(cond, truebranch)
    if (falsebranch != null) addClause(IfStatement.elseGuard, falsebranch)
  }
  
  def addClause(guard:ASTNode, stmt:ASTNode) : Unit = {
		val body : ASTNode = if (stmt == null) new BlockStatement() else stmt;
    body.setParent(this)
    if (guard != IfStatement.elseGuard) guard.setParent(this)
    cases += new IfStatementCase(guard, body)
  }
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))

  override def debugTreeChildrenFields(): Iterable[String] = Seq("cases")
  override def debugTreePropertyFields(): Iterable[String] = Seq()
}
